controllers/openshift_controller.go

```diff
@@ -135,6 +135,13 @@ func (r *KataConfigOpenShiftReconciler) Reconcile(ctx context.Context, req ctrl.
 			return ctrl.Result{}, nil
 		}
 
+		if r.kataConfig.Spec.Features.EnableTimeTravel {
```

Comment:
Instead of having featuregate in the CRD, can we have feature gates in the configmap and read it in the reconcile loop?

Reply:
That was my impression from the initial discussions, too.

Who controls the gating?  Unless we want to enable users to do it we shouldn't put it in KataConfig.

Reply:
Hi @snir911 and @bpradipt, thanks for the feedback. My thoughts:

There isn't a "standard" method for this, but since we have a CRD in place, it avoids extra complexity and additional objects.

I was concerned about managing default values and documentation with a ConfigMap (CM). However, using kubebuilder:default= simplifies default values, and we can leverage tools like swagger for documentation, keeping everything in one place.

In my testing, yes: a kataconfig is necessary, but spec.Features doesn't need to be explicitly defined; also the reconcile loop was catching the changes normally. I chose CRD to make it easier for users.

I'm open to using CM if there are benefits we are missing. My initial choice was just to avoid additional complexities.

Reply:
@pmores initiall requirements is to have a "_a user-controlled feature gating mechanism_".

Reply:
We need to make sure that whatever we put in KataConfig needs to be OK to be controlled by the user, simply because the user _owns_ their KataConfig instance.  We need to avoid repeating the kata monitor image problem.  If we put control of the features to users' hands we need to understand and accept that they e.g. might leave the features off even after we'd like them on.

Technically both CM and KataConfig allow users to set the gate values, however a CM could be undocumented and thus out of reach for users (unless they're fine with losing support). KataConfig is explicitly documented as controlled by users.

Also, changing the KataConfig CRD has been a major PITA, both because RH downstream build tools don't handle that, and because of the CRD versioning & deprecation complexity.

Reply:
I would also say that we should try to avoid using the CRD, it's very complex with respect to upgrades, versioning of the CRD, conversion webhooks etc.